src/pybind/mgr/dashboard/frontend/src/app/shared/services/module-status-guard.service.ts

```diff
@@ -0,0 +1,67 @@
+import { HttpClient } from '@angular/common/http';
+import { Injectable } from '@angular/core';
+import {
+  ActivatedRouteSnapshot,
+  CanActivate,
+  CanActivateChild,
+  Router,
+  RouterStateSnapshot
+} from '@angular/router';
+
+import 'rxjs/add/observable/of';
+import { Observable } from 'rxjs/Observable';
+
+/**
+ * This service checks if a route can be activated by executing a
+ * REST API call to '/api/<apiPath>/status'. If the returned response
+ * states that the module is not available, then the user is redirected
+ * to the specified <redirectTo> URL path.
+ *
+ * A controller implementing this endpoint should return an object of
+ * the following form:
+ * {'available': true|false, 'message': null|string}.
+ *
+ * The configuration of this guard should look like this:
+ * const routes: Routes = [
+ * {
+ *   path: 'rgw/bucket',
+ *   component: RgwBucketListComponent,
+ *   canActivate: [AuthGuardService, ModuleStatusGuardService],
+ *   data: {
+ *     moduleStatusGuardConfig: {
+ *       apiPath: 'rgw',
+ *       redirectTo: 'rgw/501'
+ *     }
+ *   }
+ * },
+ * ...
+ */
+@Injectable()
+export class ModuleStatusGuardService implements CanActivate, CanActivateChild {
+
+  constructor(private http: HttpClient,
+              private router: Router) {}
+
+  canActivate(route: ActivatedRouteSnapshot, state: RouterStateSnapshot) {
+    return this.doCheck(route);
+  }
+
+  canActivateChild(childRoute: ActivatedRouteSnapshot, state: RouterStateSnapshot) {
+    return this.doCheck(childRoute);
+  }
+
+  private doCheck(route: ActivatedRouteSnapshot) {
+    const config = route.data['moduleStatusGuardConfig'];
+    return this.http.get(`/api/${config.apiPath}/status`)
+      .map((resp: any) => {
+        if (!resp.available) {
+          this.router.navigate([config.redirectTo, resp.message]);
```

Comment:
While testing this PR, I received a 404 page:

![screenshot from 2018-05-08 16-41-53](https://user-images.githubusercontent.com/14297426/39768180-787e933a-52e0-11e8-9a94-8bfc207edecb.png)

To fix this you have to change this line to:

`this.router.navigate([config.redirectTo, resp.message || '']);`

or add add a new route to handle the case where message is `null`:

`{ path: 'rgw/501', component: Rgw501Component, canActivate: [AuthGuardService] },`